### ISSUE TITLE
Problem with handling nested projects

### ISSUE THREAD

**Issue by krzyslom** [lorenzwalthert/precommit#86]
Let's say that I have a project repo and within it I have a packages/ folder that stores the packages that are developed under this project. Every single package is also an R project.

Now I would like to have two _.pre-commit-config.yaml_ templates, one for the project scope and the other for the packages. The former will have **R package development** hooks disabled while the latter will have them enabled. The following tree represents the desired architecture.

```
. (project)
├── .git
├── .pre-commit-config.yaml (with DISABLED R package development hooks)
├── notebooks
├── packages
│   ├── packageA
│   │   ├── .pre-commit-config.yaml (with ENABLED R package development hooks)
│   │   ├── DESCRIPTION
│   │   └── ...
│   └── packageB
│       ├── .pre-commit-config.yaml (with ENABLED R package development hooks)
│       ├── DESCRIPTION
│       └── ...
├── README.md
└── ...
```

When I create a package folder inside the project directory then the package is not treated as separate git repository. If I enter the package folder (by opening the _package.Rproj_) and run `git status` then I can also see the files from the root project folder i.e. ._./../README.md_.

Running `precommit::use_precommit()` and `here::here()` from inside the package folder will result in acknowledging the _.pre-commit-conf.yaml_ file that lies within the package folder. `git commit` action also looks at the same config file. However it looks for DESCRIPTION file (as desc-in-deps hook is enabled in the package scope) inside the project root directory. This results in the following error:

```
Error: No root directory found in <my/root/project/path> or its parent directories. Root criterion: contains a file `DESCRIPTION` with contents matching `^Package:
```

thrown by _rprojroot_ package.

After the first commit failure, the package config is ignored. Now in second attempt (just hitting once more `git commit`) the one in the project root directory will be used. Now the `git commit` action will succeed because in the project root dir lies the config file that has desc-in-deps hook disabled. If I remove the yaml config from the project root directory then calling `git commit` from the package workdir sticks to looking at the config in the package workdir which again results in mentioned error.

Now if I create a totally new project from RStudio navbar, and inside that project I start another R package project with checked "Create a git repository" option and plant the yaml configs as previously, then everything will work just fine, since the precommit used within the package workdir will recognize it as the root folder since it is a separate git repository.

To sum it up -  at the moment it seems like using more than one precommit config yaml file in many subfolders within a single git repository results in a failure at first attempt, and automatic switch to the config that lies in the git repo root directory for the next attempt of `git commit`.

Is it possible to change the functionality so that upon commiting, precommit won't execute the hooks from the perspective of the yaml config in the root directory of the git repository but from the perspective of the current working directory instead i.e. subfolder with a pacakge?

**Comment by lorenzwalthert** [comment]
This is kind of a complicated question but I see you did your best to make it understandable 😄.


If I understand you correctly I think we can disentangle as follows:

1. If we make the deps-in-desc hook take the right description file, we can avoid the hook to fail and consecutively avoid for pre-commit to switch from the package root to the git root.

2. The fact that pre-commit walks up the directory tree upon a failing commit from the non-root directory is potentially unexpected. That does not have to do anything with the R package {precomit} though and had to be reported to the [upstream](https://github.com/pre-commit/pre-commit) repo. For that reason, I also cannot alter this behaviour.

3. The way you organise this project might be suboptimal. My experience is that [`git submodules`](https://git-scm.com/book/en/v2/Git-Tools-Submodules) can be a good solution to organising this type of work, or most simply, just use a flat structure like this:
```
. (project)
├── analysis
│   ├── .git
│   ├── notebooks
│   ├── .pre-commit-config.yaml (with DISABLED R package development hooks)
├── tools
│   ├── packageA
│   │   ├── .git
│   │   ├── .pre-commit-config.yaml (with ENABLED R package development hooks)
│   │   ├── DESCRIPTION
│   │   └── ...
│   ├── packageB
│   │   ├── .git
│   │   ├── .pre-commit-config.yaml (with ENABLED R package development hooks)
│   │   ├── DESCRIPTION
│   │   └── ...
```

An alternative structure is to use a bitbucket-like structure with team / project / repo. Have a look at https://github.com/lorenzwalthert/teamtools, but note it's not under active development.

**Comment by krzyslom** [comment]
Thank you for the quick reply and a tip with git submodules.

The walk up the directory tree upon a failing commit clouds the core of the issue and indeed could be a separate ticket related to the upstream project as you mentioned in 2.

I've tried to simplify the core of the issue and break it into steps.

A)
1. Create an R project called **project** that is a git repository
1. Open the ~/project/project.Rproj
1. Create an R package project called **package** that is not a separate git repo
1. The structure at the moment looks like this

```
. (project)
├── .git
├── .gitignore
├── project.Rproj
└── package
    ├── DESCRIPTION
    ├── package.Rproj
    ├── README.md
    └── ...
```

B) 
1. Open the ~/project/package.Rproj
1. Call `precommit::use_precommit()` having ~/project/package as a workdir
1. Enable the desc-in-deps hook in freshly opened one and only yaml config
1. `git add` and `git commit` results in the following error

```
No .pre-commit-config.yaml file was found
- To temporarily silence this, run `PRE_COMMIT_ALLOW_NO_CONFIG=1 git ...`
- To permanently silence this, install pre-commit with the --allow-missing-config option
- To uninstall pre-commit run `pre-commit uninstall`
```

So the precommit fails to recognize that I actually have a yaml config in my workdir. I guess since my workdir is not the main project/ folder (the root of the git repo).

C)
1. Once again call `precommit::use_precommit()`.
1. And again call `git commit`.

Surprisingly now the hooks are executed (and nothing has been explicitly changed by the developer since the previous error. We are just calling the same function and the same command the second time). However this action results in the following error:

```
Error: No root directory found in ~/project or its parent directories. Root criterion: contains a file `DESCRIPTION` with contents matching `^Package: `
Execution halted
```

This is very weird, since precommit knows that the desc-in-deps hook is enabled and the yaml config can be found in the workdir however it does not act from the perspective of this config file.

Adding second yaml config to the project/ folder with disabled desc-in-deps hook brings another level of confusion to the topic. 

The ideal(?) situation would be if the precommit executed hooks from the perspective of the current working directory.

I hope this clarifies things a bit.

**Comment by lorenzwalthert** [comment]
Ok, thanks. To be honest, I don't think this workflow is encouraged by pre-commit.com. Because the assumption that your config file is in the git root and this is also the project root has a lot of implications in various places in the framework I assume so I am not sure if the maintainer of pre-commit is willing to give that up for limited gain. Feel free to open an issue in the upstream repo and link it here.

The inconsistency you observe is probably because the config is taken from your working directory, but the hook is executed from the git root as the working directory. This will most likely also affect other hooks because all your files are not in the places relative to the git root from where they are usually, e.g. `R/`. I'd advise to use the flat structure I proposed above. Otherwise, just disable the hooks that cause problems.
